### DISCUSSION TITLE
Further project structure

### DISCUSSION THREAD

**Discussion by louishamelers** [CapsCollective/raylib-cpp-starter#9] [Q&A] [answered]
Beginner over here! I'm wondering if the makefile supports multiple classes in sub directories. For example, if I want to make a SpriteController class, where would I put the header and cpp files for it to be compiled automatically? My current guess would be like this:
![image](https://user-images.githubusercontent.com/25366619/129862800-4c444893-e5ed-47a7-993b-03947cb899f4.png)
But that doesn't work :(

Any help would be greatly appreciated!

**Comment by jonjondev** [comment, marked as answer]
Hi @louishamelers!

Thanks for the question because we've had a similar questions on header placement a few times now. Forgive me because I'm going to provide a long-winded explanation for clarity.

If you take a look at the `Makefile`, you'll see the following lines in the first paragraph or so of code:
```Makefile
sources := $(call rwildcard,src/,*.cpp)
...
compileFlags := -std=c++17 -I include
```

The `sources` macro line will essentially generate a space-separated list of filepaths to all files ending in `.cpp`, inside the folder `src` (recursively) to be used as "sources". Thus, all of these will be compiled into objects (`.o` files) that are then linked into the executable.

**The headers are never considered as translation units**, meaning that they are pretty much orthogonal to the compilation process. They only get seen by the compiler if you mention them in an `#include` directive. You can pretty much think of an include directive as copy-pasting the contents of the `.h` file into the top of the `.cpp` file that references it.

Another interesting thing about those `#include` directives is that they accept relative paths. So with the file structure:

```
src/
  |_ headers/
  |   |_ SpriteController.h
  |_ sources/
  |   |_ SpriteController.cpp
  |_ main.cpp
```

Everything should compile just fine as long as you mention in your `main.cpp` and `SpriteController.cpp` that you'd like to `#include "headers/SpriteController.h"` and `#include "../headers/SpriteController.h"` respectively. If that's the case, let's take it further - the following would also work:

```
headers/
  |_ SpriteController.h
src/
  |_ main.cpp
  |_ SpriteController.cpp
```

And the only change needed would be to tell `main.cpp` it's now another directory up (`../`)!

Also nothing stopping you from putting them next to each other:

```
src/
  |_ main.cpp
  |_ SpriteController.cpp
  |_ SpriteController.h
```

The main point is that the **compiler needs to know which source files to compile** (which we've designated as anything in `src`) **and each `.cpp` file needs to know where the headers are that it includes**. That is the only rule here.

Additionally, I started this reply by mentioning another line from the `Makefile`, the `compileFlags` macro. I brought it up because it references `include`, a directory **full of `.h` files!** However, the `.h` files it contains are those used by `raylib` and `raylib-cpp`, the libraries we use.

You can _technically_ put files in there, and reference them without relative path (as so: `#include <SpriteController.h>`) because we set the `include` directory as a pool of globally accessible includes using `-I include`, but it's not very nice and a little rude.

If you **did** want that kind of syntax for includes, but you didn't want to be rude, you could just... add a cheeky little `-I src/headers` to the `compileFlags` and there you go!

Getting to the point, the answer here is that there are tens of approaches to how you setup **your** build system and using Make. The template is incredibly flexible so if you want to take a look into how you'd go about making adjustments, [just check out our doc explaining the Makefile line-by-line](https://github.com/CapsCollective/raylib-cpp-starter/blob/main/docs/MakefileExplanation.md).

Additionally if you want some inspiration on how configurable the structure of the project is, you can take a look at [our test-pilot project](https://github.com/CapsCollective/a-dark-discomfort) which, starting from this template has evolved to suit the needs of a game engine, and which we continue to use as a proving ground for the robustness of the template.

Let me know if the response appropriately covers all your questions (and probably a lot more you didn't ask). I also sincerely hope that I am correct in my assumption of why your implementation didn't compile and it's not some weird Windows issue (as Windows often supplies).

**Reply by louishamelers** [reply]
Thanks for the elaborate answer, you are an absolute hero!
The culprits ended up being the static references like you said. And thank you so much for the makefile documentation reference and the test pilot project, I will check it out!
